### PR TITLE
[RUNTIME][DSO] Improve TVMBackendPackedCFunc to allow return val

### PR DESCRIPTION
**Pull request by tqchen**
Previously the signature of LibraryModule's PackedFunc does not support return value.
This wasn't a limitation for our current usecase but could become one
as we start to generate more interesting functions.

This feature will start to get interesting as we move towards unified
object protocol and start to pass object around.
This PR enhances the function signature to allow return values.

We also created two macros TVM_DLL_EXPORT_PACKED_FUNC and TVM_DLL_EXPORT_TYPED_FUNC
to allow manual creation of functions that can be loaded by a LibraryModule.

Examples are added in apps/dso_plugin_module.
The change to TVMBackendPackedCFunc is backward compatible,
as previous function will simply ignore the return value field.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @comaniac @zhiics @wweic @yzhliu

**Comment by comaniac** [comment]
Thanks for the PR. The macros are indeed helpful for customized runtimes and can reduce the developer efforts. I'll file another PR to make use of them in [this example](https://github.com/apache/incubator-tvm/blob/master/src/runtime/contrib/example_ext_runtime/example_ext_runtime.cc) after this one have been merged. 

The change are LGTM. Since I may not be aware of some concerns in backend APIs, I am leaving it for others.

File: include/tvm/runtime/packed_func.h [hunk 1]
```diff
@@ -877,6 +894,104 @@ class TVMRetValue : public TVMPODValue_ {
   }
 };
 
+/*!
+ * \brief Export a function with the PackedFunc signature
+ *        as a PackedFunc that can loaded by LibraryModule.
```
Comment (zhiics) [inline include/tvm/runtime/packed_func.h]:
```suggestion
 *        as a PackedFunc that can be loaded by LibraryModule.
```

File: include/tvm/runtime/packed_func.h [hunk 2]
```diff
@@ -877,6 +894,104 @@ class TVMRetValue : public TVMPODValue_ {
...
+    } catch (const ::std::runtime_error& _except_) {                    \
+      TVMAPISetLastError(_except_.what());                              \
+      return -1;                                                        \
+    }                                                                   \
+  }                                                                     \
+  }
+
+/*!
+ * \brief Export typed function as a PackedFunc
+ *        that can loaded by LibraryModule.
```
Comment (zhiics) [inline include/tvm/runtime/packed_func.h]:
```suggestion
 *        that can loaded be by LibraryModule.
```

File: apps/dso_plugin_module/test_plugin_module.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+def test_plugin_module():
+    curr_path = os.path.dirname(os.path.abspath(os.path.expanduser(__file__)))
+    mod = tvm.module.load(os.path.join(curr_path, "lib", "plugin_module.so"))
+    # NOTE: we need to make sure all managed resources returned
+    # from mod get destructed before mod get unloaded.
+    #
+    # Failure mode we want to prevent from:
+    # We retain an object X whose destructor is within mod.
+    # The program will segfault if X get destructed after mod,
+    # because the destructor function has already been unloadd.
```
Comment (wweic) [inline apps/dso_plugin_module/test_plugin_module.py]:
```suggestion
    # because the destructor function has already been unloaded.
```

**Comment by tqchen** [comment]
Thanks @wweic @ZihengJiang @zhiics

**Pull request merged**